### ISSUE TITLE
"Sharing outputs between projects" > doesn't work, documentation incomplete?

### ISSUE THREAD

**Issue by aSemy** [gradle/gradle#18846]
I've tried following the guide https://docs.gradle.org/7.2/userguide/cross_project_publications.html but it doesn't work. I can't figure out how to create or share files.

I'm generating some Typescript, Kotlin, and Java files with `protoc`. I've written a Gradle task to do this. This is happening in one subproject - `data-model`.

I've tried creating a configuration 

##### producer

```kotlin
  val tsProto by configurations.creating {
    isCanBeResolved = false
    isCanBeConsumed = true
  }
```

And I've got a task that produces Typescript files

```kotlin
  val protobufTypescript by tasks.creating(ProtobufCompileTask::class) {
    description = "proto2typescript"
    protoFile.set(file("$projectDir/src/proto/Record.proto"))

    cliArgs.add("--ts_proto_out=${project.mkdir("$temporaryDir/typescript").canonicalPath}")
  }
```

I add the task output to an artifact 

```kotlin
  artifacts {
    add(tsProto.name, protobufTypescript.outputDir) {
      builtBy(protobufTypescript)
    }
  }
```

##### consumer

Now in the consumer I've tried setting up a dependency, but it doesn't receive anything.

```kotlin
val tsProto: Configuration by configurations.creating {
  isCanBeConsumed = false
  isCanBeResolved = true
}

dependencies {
  tsProto(projects.modules.dataModel.dependencyProject.configurations.getByName("tsProto"))
}

// no output
tsProto.resolve().forEach {
  logger.lifecycle("---- tsProto file in mod project: $it")

}

// no output
tsProto.artifacts.files.files.forEach {
  logger.lifecycle("tsProto file in mod project: $it")
}
```

### Expected Behavior

Code is shared between subprojects 

### Current Behavior

The documentation doesn't make it clear how to do this. It's impossible to figure out. 

### Context

I want to generate code in one subproject, and use it in other subprojects.

This is quite frustrating because the documentation says "don't reference other project tasks directly", but when I do that it works perfectly first time.

```kotlin

projects.modules.factorioEventsDataModel.dependencyProject.tasks.named("protobufTypescript")
  .get().outputs.files.forEach {
  logger.lifecycle("!!!tsProto file in mod project: $it")
}
// output: !!!tsProto file in mod project: ...data-model\build\tmp\protobufTypescript

```

### Your Environment

Gradle 7.2, IntelliJ IDEA 2021.2.3, jdk 11, Windows 10, Kotlin: 1.5.31

**Comment by vlsi** [comment]
@aSemy , you are doing it all wrong (tm) :)

In fact, you almost got it right :)

In consumer, you should not mention the configuration name.
The idea is it is not the consumer's business how the configuration is named, and the names of the configurations are private to the project.

The consumer should declare the dependency as follows:

```kotlin
dependencies {
  tsProto(projects.modules.dataModel.dependencyProject)
// or even
// tsProto("com.acme:some-dependency:1.0")) 
}
```

However, you might ask: "how Gradle would know which artifacts does that mean?"
The answer is "it uses attributes to classify the artifacts".

In other words:
Producer: "ok I have these artifacts for 'typescript definitions', these artifacts for 'documentation', these artifacts for 'code coverage', these artifacts for 'source code',..."
Consumer: "dear producer, would you please provide artifacts with typescript definitions?"

----

So the missing bits are:

Producer:

(see https://github.com/gradle/gradle/pull/16627 )

```
inline fun <reified T: Named> AttributeContainer.attribute(attr: Attribute<T>, value: String) =
    attribute(attr, objects.named<T>(value))

  val tsProto by configurations.creating {
    isCanBeResolved = false
    isCanBeConsumed = true
    attributes {
        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, "typescript-definitions")
    }
  }
```

Consumer:

```
val tsProto: Configuration by configurations.creating {
  isCanBeConsumed = false
  isCanBeResolved = true
    attributes {
        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, "typescript-definitions")
    }
}
```

The nice consequence is that in consumer you declare "dependencies" in a single place and make `tsProto`-like configurations to extend from it.
See https://github.com/jjohannes/idiomatic-gradle/blob/d8677b0242e360a1c5a0a349ada55efb078995ec/aggregation/package-server/build.gradle.kts#L7 and https://github.com/jjohannes/idiomatic-gradle/blob/d8677b0242e360a1c5a0a349ada55efb078995ec/build-logic/packaging/src/main/kotlin/com.example.jacoco-aggregation.gradle.kts#L8-L16

Of course, choosing the attributes is somewhat hard task, and you might find that pre-existing attributes make sense for you (e.g. LibraryElements, Usage, Category, Bundling, OperatingSystemFamily, MachineArchitecture, etc)

**Comment by ljacomet** [comment]
Thanks Vladimir for taking the time to answer.

@aSemy Let us know if this resolves your attempt and if you have improvement suggestion for the documentation. Otherwise I will close this issue in the coming days as there is no change needed on the Gradle side.

**Comment by aSemy** [comment]
@vlsi Thank you very much for the detailed response

> In consumer, you should not mention the configuration name. The idea is it is not the consumer's business how the configuration is named, and the names of the configurations are private to the project.

That makes sense to me, but I'm very confused because in example 4 the configuration that is produced (`instrumentedJars`) *is* named in the consumer. Is this wrong?

```kotlin
dependencies {
    instrumentedClasspath(project(mapOf(
        "path" to ":producer",
        "configuration" to "instrumentedJars")))
}
```

In your example doesn't the consumer *also* have to know how the producer has named their configuration? I see `"typescript-definitions"` twice, in consumer and producer. What's the benefit of doing this, but also flipping some boolean flags over? Why is that better than getting the output of a task? Both approaches seem fragile to me, but the difference is that with the 'task' approach it's less code, so it's easier to figure out what's going on, and there's fewer places it could go wrong. 

Plus I don't have to figure out what to do with these whole attribute naming thing! That is also very unclear to me - I really haven't got enough information to understand the whole process, let alone make a choice about something that apparently is very difficult. I would really feel better if the Gradle code had a basic template to work from. 

---

@ljacomet Feel free to close this ticket if there's nothing wrong :)

The documentation could be improved by a complete example. At the moment it's split out over several sections that feel out of order. For example, `instrumentedClasspath` is defined *after* it's used in code. 

It's also tough to understand because the language used in the documentation doesn't match the code. 'Variants', 'outputs' - the code doesn't mention these words so it's difficult for me to make a connection. 'Configurations' is used a lot in code and the docs, but that's a really generic word that to me doesn't say "files I want to share with another project", but instead "here's how I want to configure my project's behaviour".

I know it's not very constructive feedback, but I get really quite lost in the 'Variant-aware sharing of artifacts between projects' section. I'm not sure what to suggest. My understanding was that 'variant' was more about when code is compiled into an executable for Mac or Windows or Linux - three variants to be shared, which isn't applicable in my case, so I don't need it? It's tough to understand. 

Also it would be nice to have some helper functions to make resource sharing easier. At the moment the 'anti-pattern' actually works, but the 'correct' way requires a lot more work that is finickety to set up. to me that's less of an anti-pattern, and more of a convenient way of working. It might be fragile, but at least it works.

**Comment by tbroyer** [comment]
Speaking of improvements to the docs, in the Kotlin DSL examples, the project dependency could use the Kotlin DSL extension:

```kotlin
dependencies {
    instrumentedClasspath(project(path = ":producer", configuration = "instrumentedJars"))
}
```
rather than the current
```kotlin
dependencies {
    instrumentedClasspath(project(mapOf(
        "path" to ":producer",
        "configuration" to "instrumentedJars")))
}
```

**Comment by vlsi** [comment]
>Is this wrong?

```groovy
dependencies {
    instrumentedClasspath(project(mapOf(
        "path" to ":producer",
        "configuration" to "instrumentedJars")))
}
```

It is wrong for exactly the reason you mention: "consumer also have to know how the producer has named their configuration".

> Plus I don't have to figure out what to do with these whole attribute naming thing! 

The attributes are not that hard.
The producer uses attributes to describe the properties (e.g. `I produce typescript-definitions`), and the consumer declares what they need (e.g. `I need typescript-definitions`). It is not that hard, however, I would agree, the first time I faced Gradle attributes I was puzzled.

The example I suggested in https://github.com/gradle/gradle/issues/18846#issuecomment-960582717 should be just enough for you, and it contains just a couple of lines for both the producer and the consumer.

---

@ljacomet , I would prefer mentioning attribute-based sharing first, and list "configuration-name based" approach somewhat later as a workaround "if you can't use attributes for some reason or need an urgent fix and you are eager to cut corners". WDYT?

The current DOC starts with a workaround-like approach, so the readers get lost.

**Comment by aSemy** [comment]
I'm coming back to this problem today because I've learned a lot about Gradle since first writing this issue, but I still find sharing files between files incredibly confusing and frustrating.

* The API naming is confusing. 
   * `Configuration` doesn't match the common usage of the word configuration, aka 'settings' or 'properties', and doesn't suggest "use this to share files between projects'. 
   * `isCanBeResolved` and `isCanBeConsumed` seem like synonyms, so I always mix them up.
* Files get randomly added to a configuration #23776 
* Sometimes Configurations just don't seem to work #6881 
* I've tried setting up file sharing, but when it just doesn't work I don't know where to turn. Debugging where files go and attribute selection and Configuration inheritance is really difficult because of the lazy API - I want to be able to add a debug action to a Provider so I can see what the values are. I want to follow where files go. I want to find out why the Configurations and attributes that I've made aren't working.
